ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/BalanceAttackMitigationForkChoiceTrigger.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.forkchoice;
+
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+/**
+ * A fork choice trigger that implements the LMD GHOST Balance Attack mitigation from HF1.
+ *
+ * @see <a href="https://hackmd.io/nyKbwIluQlWBH6sli4tJlA?view">Implementation plan</a>
+ */
+class BalanceAttackMitigationForkChoiceTrigger implements ForkChoiceTrigger {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private final AtomicReference<ForkChoiceUpdate> latestCompletedForkChoice =
+      new AtomicReference<>();
+  private final ForkChoice forkChoice;
+
+  BalanceAttackMitigationForkChoiceTrigger(final ForkChoice forkChoice) {
+    this.forkChoice = forkChoice;
+  }
+
+  @Override
+  public void onSlotStartedWhileSyncing(final UInt64 nodeSlot) {
+    processHead(nodeSlot);
+  }
+
+  @Override
+  public void onSlotStarted(final UInt64 nodeSlot) {
+    processHead(nodeSlot);
+  }
+
+  @Override
+  public void onAttestationsDueForSlot(final UInt64 nodeSlot) {}
+
+  @Override
+  public SafeFuture<Void> prepareForBlockProduction(final UInt64 slot) {
+    final ForkChoiceUpdate forkChoiceUpdate = processHead(slot);
+    if (forkChoiceUpdate.nodeSlot.isGreaterThan(slot)) {
```

Comment:
In this case, it seems like we're not actually guaranteed to have processed fork choice up to `slot`.  For example, say we're wanting fork choice to run at slot 10 and processing for slot 10 failed.  If slot 11 is queued, we'll carry on assuming we've processed slot 10.  Maybe that's fine since we're already late at that point anyway?

Reply:
hmm, good point.  I suspect failing to run fork choice is a pretty massive problem but it will make sense most of the the time to operate as if it had worked.  Best to be consistent about that though so I'll add an `exceptionally` block to log any errors so that the returned future always completes.  We don't want to fail block production because something went wrong in fork choice - better to produce a block and just hope we're on the right fork.